### PR TITLE
feat: support parsing `satisfies` operators

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## PR Checklist

- [x] Addresses an existing open issue: partial of #5688
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Based on #5716
Adds parsing support for [`satisfies` operator](https://devblogs.microsoft.com/typescript/announcing-typescript-4-9-beta/#the-satisfies-operator).

### CODE REVIEW

File: packages/typescript-estree/tests/ast-alignment/fixtures-to-test.ts [hunk 1]
```diff
@@ -477,6 +477,11 @@ tester.addFixturePatternConfig('typescript/expressions', {
      * @see https://github.com/babel/babel/issues/14613
      */
     'instantiation-expression',
+    /**
+     * TS 4.9 `satisfies` operator has not been implemented in Babel yet.
+     * @see https://github.com/babel/babel/pull/14211
```
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/tests/ast-alignment/fixtures-to-test.ts:482]: That PR is now merged. 🎉

File: packages/shared-fixtures/fixtures/typescript/expressions/satisfies-expression.src.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+x satisfies T;
+x < y satisfies boolean; // (x < y) satisfies boolean;
+x === 1 satisfies number; // x === (1 satisfies number);
+x satisfies any satisfies T;
+
+const t2 = { a: 1, b: 1 } satisfies I1; // Error
+const t3 = { } satisfies I1; // Error
+const t4: T1 = { a: "a" } satisfies T1; // Ok
+const t5 = (m => m.substring(0)) satisfies T2; // Ok
+const t6 = [1, 2] satisfies [number, number];
```
Comment (JoshuaKGoldberg) [inline packages/shared-fixtures/fixtures/typescript/expressions/satisfies-expression.src.ts:10]: Just making sure inline function types are supported too...

```suggestion
const t6 = [1, 2] satisfies [number, number];
const t7 = (async (context) => context.value) satisfies (context: { value: string }) => Promise<string>;
```

File: package.json [hunk 3]
```diff
@@ -104,10 +104,10 @@
     "tmp": "^0.2.1",
     "ts-node": "^10.7.0",
     "tslint": "^6.1.3",
-    "typescript": ">=3.3.1 <4.9.0"
+    "typescript": ">=3.3.1 <4.8.3 || 4.9.1-beta"
   },
   "resolutions": {
-    "typescript": "~4.8.3",
+    "typescript": "~4.9.1-beta",
```
Comment (JoshuaKGoldberg) [inline package.json]:
```suggestion
    "typescript": "~4.9.3",
```

File: packages/typescript-estree/tests/ast-alignment/fixtures-to-test.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/typescript-estree/tests/ast-alignment/fixtures-to-test.ts:482]: Mentioned this in #6027.

**Review by bradzacher — APPROVED** [review]
thanks for doing this!

thankss

For reference, changes I made prior to merging:
- update babel packages so it supports the satisfies operator
- converted the fixture test from a "shared fixtures" test to a new and shiny "ast-spec" test.
    - this is the preferred way of doing these ast tests going forward as we can more easily split them, compare with babel etc

**Pull request merged**